OAuth: Added support for role mapping for GitLab accounts.

splattael opened the pull request:
Allow role mapping for GitLab accounts.

Example:

```ini
[auth.gitlab]
role_attribute_path = is_admin && 'Admin' || 'Viewer'
```

## Tasks

- [ ] Add tests
- [ ] Add documentation (see https://grafana.com/docs/grafana/latest/auth/okta/#map-roles)

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR adds support for `role_attribute_path` JMESPath to allow role mapping for GitLab social login.

**Which issue(s) this PR fixes**:

Fixes #28892

**Special notes for your reviewer**:

splattael committed:
Support `role_attribute_path` for GitLab OAuth

Allow role mapping for GitLab accounts.

Example:

  [auth.gitlab]
  role_attribute_path = is_admin && 'Admin' || 'Viewer'

splattael committed:
Support `role_attribute_path` for GitLab OAuth

Allow role mapping for GitLab accounts.

Example:

  [auth.gitlab]
  role_attribute_path = is_admin && 'Admin' || 'Viewer'

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=30025) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=30025) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=30025) it.</sub>

splattael commented:
@papagian Despite the missing tests and docs to do you think the PR is worth keeping? :thinking: :sweat_smile:

hobti01 commented:
Are tests/docs blocking this from being accepted?  We would be very interested in this functionality.

MCrank commented:
I'd like to second what @hobti01 says as this functionality would be very welcome for us along with several others based on this issue. 
https://github.com/grafana/grafana/issues/28892. 

Also here on the Gitlab side of the house
https://gitlab.com/gitlab-org/gitlab/-/issues/31125

marefr commented:
@splattael thanks. Sorry for the late respons, we managed to not notice this one until now. This looks definitely like something we can accept. If you can sync with main branch, sign the cla and add documentation we can review/merge this. Adding tests would be nice as well, but maybe tricky since there are no existing tests for gitlab, but looking at generic oauth tests might give you an idea - let us know if you're still planning to work on this and/or need help.

henrysachs commented:
Hey There, any update on this. We would like to use this feature. If I notice it correctly its only missing the signed CLA? :)

splattael commented:
@henrysachs 

> If I notice it correctly its only missing the signed CLA? :)

Yes but also

> If you can sync with main branch, sign the cla and **add documentation we can review/merge this**.

(_highlights by me_)

:sweat_smile: 

I can start with signing the CLA and rebasing :grin: 

@henrysachs Any kind of contribution is very welcome :heart:

henrysachs commented:
I can write some kind of docs for this. Will have a look onto that. :)

splattael commented:
@henrysachs That'd be awesome :tada: Thanks for offering help :heart: Happy to contribute on that and move this finally forward :sweat_smile:

henrysachs commented:
Hey @splattael what is you opinion on querying the groups also via JMESPath. I can imagine queries based on groups users are in. I think this could be done pretty easily and would be a great benefit.

henrysachs commented:
and what is your preferred way to get my changes? I can commit them right onto your branch if you want to. I wrote some docs inspired by the octa ones :)

splattael commented:
@henrysachs Good suggestion! Querying GitLab groups via JMESPath could be very useful indeed :+1: 

I'd love to keep the PRs focused and small so maybe we make separate PR? :thinking:

splattael commented:
> and what is your preferred way to get my changes? I can commit them right onto your branch if you want to. I wrote some docs inspired by the octa ones :)

I've sent your an collaboration invitation for my fork https://github.com/splattael/grafana. Would this work for you?

henrysachs commented:
If you don't mind I would like to try and implement this. I love getting OS contributions 😅

henrysachs commented:
Or tbh I'm already quite far with implementing it 😅

splattael commented:
@henrysachs Sure, go for it! Please feel free to ping me if there's something to test :grin: 

My golang fu is limited as my help is :sweat_smile:

marefr commented:
Am I understanding correctly that more work is planned for this PR/branch? Regarding "Querying GitLab groups via JMESPath" that's definitely its own feature/PR.

henrysachs commented:
As it were just more attributes to query i thought it would be the same but we can also split this. :)

henrysachs commented:
So i have an first implementation of this ready. I really would like to keep this in the same Branch/PR but happy to discuss this. I'm just using the role_attribute_path against the groups result from gitlab.

splattael commented:
> As it were just more attributes to query i thought it would be the same but we can also split this. :)

@henrysachs I'd prefer to split this. If you want to rely on `role_attribute_path` for groups, you could branch from this very feature branch :grinning: 

> Am I understanding correctly that more work is planned for this PR/branch?

@marefr Yes! @henrysachs offered help regarding docs. I've signed CLA and rebased against latest `master` (fixing some lint errors). The only thing missing are tests :sweat_smile: This could be a follow-up PR though? :thinking: 

> Regarding "Querying GitLab groups via JMESPath" that's definitely its own feature/PR.

:+1:

henrysachs commented:
Okay so should I commit my docs regarding the query? Or should I just do them in the other pr? :)

henrysachs committed:
Merge branch 'gitlab-auth-role-attribute-path' of github.com:splattael/grafana into gitlab-auth-role-attribute-path

henrysachs committed:
docs: add docs for role_attribute_path

henrysachs commented:
just committet the docs and will open a separate pr for the group mapping.

splattael commented on docs/sources/auth/gitlab.md:
````diff
@@ -103,13 +103,27 @@ enabled = true
+role_attribute_path = is_admin && 'Admin' || 'Viewer'
+```
+
+This would allow you every Gitlab Admin to be an Admin in Grafana.
````
```suggestion
This would allow you every GitLab Admin to be an Admin in Grafana.
```

splattael commented on docs/sources/auth/gitlab.md:
```diff
@@ -103,13 +103,27 @@ enabled = true
 
+### Map roles
+
+Grafana can attempt to do role mapping through Gitlab OAuth. In order to achieve this, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option.
```
```suggestion
Grafana can attempt to do role mapping through GitLab OAuth. In order to achieve this, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option.
```

splattael commented on docs/sources/auth/gitlab.md:
```diff
@@ -103,13 +103,27 @@ enabled = true
+
+Grafana can attempt to do role mapping through Gitlab OAuth. In order to achieve this, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option.
+
+Grafana uses JSON obtained from querying the `/api/v4/user` endpoint for the path lookup. The result after evaluating the `role_attribute_path` JMESPath expression needs to be a valid Grafana role, i.e. `Viewer`, `Editor` or `Admin`. Refer to [Organization roles]({{< relref "../permissions/organization_roles.md" >}}) for more information about roles and permissions in Grafana.
```
```suggestion
Grafana uses JSON obtained from querying GitLab's API [`/api/v4/user`](https://docs.gitlab.com/ee/api/users.html#list-current-user-for-normal-users) endpoint for the path lookup. The result after evaluating the `role_attribute_path` JMESPath expression needs to be a valid Grafana role, i.e. `Viewer`, `Editor` or `Admin`. Refer to [Organization roles]({{< relref "../permissions/organization_roles.md" >}}) for more information about roles and permissions in Grafana.
```

splattael commented on docs/sources/auth/gitlab.md:
````diff
@@ -103,13 +103,27 @@ enabled = true
+An example Query could look like the following:
+
+```bash
+role_attribute_path = is_admin && 'Admin' || 'Viewer'
````
**Suggestion** What are your thoughts on copying this great example into the first `[auth.gitlab]` example above https://github.com/grafana/grafana/pull/30025/files#diff-1d2cd961688172e12c62c2902ccbe4a45e2f990d82fda8e48516099b58b2b9e5L53? :thinking: 

**Question** Do you think this JMESPath query could be the default value in the default `[auth.gitlab]` configuration sample? :shrug: It's also a question for other reviewers I feel.

henrysachs replied:
Yeah, sounds great to me! Probably this is the most wanted query.

splattael reviewed (commented):
@henrysachs Great docs, thank you for adding them :bow: 

I left some minor suggestions for your consideration and a question for you but also other reviewers :sweat_smile:

henrysachs commented:
I read through them on mobile and will commit them once I'm back home. 😊

henrysachs committed:
Apply suggestions from code review

Co-authored-by: Peter Leitzen <splattael@users.noreply.github.com>

henrysachs committed:
docs: update example

example should suggest a full configuration

henrysachs commented:
So I implemented all your suggestions but moved the role_attribute query to the last example as it represents the "full blown" configuration that a user may want. I also used that one when setting up authentication for my work instance. :) 
Also tried to use the uppercase L on GitLab :D

henrysachs commented:
Also I'll do my other changes as a follow up PR, because I can just fork the main repo and don't have to do a fork of forks. Or am I wrong?

marefr commented on docs/sources/auth/gitlab.md:
````diff
@@ -26,10 +26,10 @@ instance, if you access Grafana at `http://203.0.113.31:3000`, you should use
 ```
 
-Finally, select _read_api_ as the _Scope_ and submit the form. Note that if you're
+Finally, select _read_api_as the_Scope_and submit the form. Note that if you're
````
```suggestion
Finally, select _read_api as the _Scope_ and submit the form. Note that if you're
```

marefr commented on docs/sources/auth/gitlab.md:
```diff
@@ -26,10 +26,10 @@ instance, if you access Grafana at `http://203.0.113.31:3000`, you should use
 not going to use GitLab groups for authorization (i.e. not setting
-`allowed_groups`, see below), you can select _read_user_ instead of _read_api_ as
-the _Scope_, thus giving a more restricted access to your GitLab API.
+`allowed_groups`, see below), you can select_read_user_ instead of _read_api_as
+the_Scope_, thus giving a more restricted access to your GitLab API.
```
```suggestion
`allowed_groups`, see below), you can select _read_user_ instead of _read_api as
the _Scope_, thus giving a more restricted access to your GitLab API.
```

marefr reviewed (changes requested):
Some minor docs suggestions. But looks good in general

henrysachs committed:
Apply suggestions from code review

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

marefr reviewed (approved):
LGTM

marefr commented:
I've asked technical writers to review this as well

splattael commented:
@henrysachs Thanks for updating the docs again :bow: 

@marefr Thanks for review and approval :heart: 

CI complains:

> continuous-integration/drone/pr — Build was killed

Is this something we need to address or is it fine? :sweat_smile: :crossed_fingers:

marefr commented:
I restarted the build. See if that resolves things :pray:

henrysachs commented:
everythings green! nice. Any updates from the technical writers? :)

GrafanaWriter commented on docs/sources/auth/gitlab.md:
````diff
@@ -26,9 +26,9 @@ instance, if you access Grafana at `http://203.0.113.31:3000`, you should use
 ```
 
-Finally, select _read_api_ as the _Scope_ and submit the form. Note that if you're
+Finally, select _read_api as the _Scope_ and submit the form. Note that if you're
````
I suspect we want to keep `read_api` and `Scope` as italics here:
```suggestion
Finally, select _read_api_ as the _Scope_ and submit the form. Note that if you're
```

GrafanaWriter commented on docs/sources/auth/gitlab.md:
```diff
@@ -26,9 +26,9 @@ instance, if you access Grafana at `http://203.0.113.31:3000`, you should use
+Finally, select _read_api as the _Scope_ and submit the form. Note that if you're
 not going to use GitLab groups for authorization (i.e. not setting
-`allowed_groups`, see below), you can select _read_user_ instead of _read_api_ as
+`allowed_groups`, see below), you can select _read_user_ instead of _read_api as
```
```suggestion
`allowed_groups`, see below), you can select _read_user_ instead of _read_api_ as
```

GrafanaWriter commented on docs/sources/auth/gitlab.md:
```diff
@@ -95,21 +95,36 @@ characters. Make sure you always use the group or subgroup name as it appears
 in the URL of the group or subgroup.
 
 Here's a complete example with `allow_sign_up` enabled, and access limited to
```
```suggestion
Here's a complete example with `allow_sign_up` enabled, with access limited to
```

GrafanaWriter commented on docs/sources/auth/gitlab.md:
```diff
@@ -95,21 +95,36 @@ characters. Make sure you always use the group or subgroup name as it appears
 
 Here's a complete example with `allow_sign_up` enabled, and access limited to
-the `example` and `foo/bar` groups:
+the `example` and `foo/bar` groups it also promotes all GitLab Admins to Grafana Admins:
```
```suggestion
the `example` and `foo/bar` groups. The example also promotes all GitLab Admins to Grafana Admins:
```

achatterjee-grafana commented:
> everythings green! nice. Any updates from the technical writers? :)

On it.

GrafanaWriter commented on docs/sources/auth/gitlab.md:
```diff
@@ -95,21 +95,36 @@ characters. Make sure you always use the group or subgroup name as it appears
 
+### Map roles
+
+Grafana can attempt to do role mapping through GitLab OAuth. In order to achieve this, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option.
```
```suggestion
Grafana can attempt role mapping through GitLab OAuth. To do so, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified using the `role_attribute_path` configuration option.
```

GrafanaWriter commented on docs/sources/auth/gitlab.md:
```diff
@@ -95,21 +95,36 @@ characters. Make sure you always use the group or subgroup name as it appears
+
+Grafana can attempt to do role mapping through GitLab OAuth. In order to achieve this, Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option.
+
+Grafana uses JSON obtained from querying GitLab's API [`/api/v4/user`](https://docs.gitlab.com/ee/api/users.html#list-current-user-for-normal-users) endpoint for the path lookup. The result after evaluating the `role_attribute_path` JMESPath expression needs to be a valid Grafana role, i.e. `Viewer`, `Editor` or `Admin`. Refer to [Organization roles]({{< relref "../permissions/organization_roles. ...
```
```suggestion
Grafana uses JSON obtained from querying GitLab's API [`/api/v4/user`](https://docs.gitlab.com/ee/api/users.html#list-current-user-for-normal-users) endpoint for the path lookup. The result after evaluating the `role_attribute_path` JMESPath expression needs to be a valid Grafana role; `Viewer`, `Editor`, or `Admin`. Refer to [Organization roles]({{< relref "../permissions/organization_roles.md" >}}) for more information about roles and permissions in Grafana.
```

achatterjee-grafana replied:
```suggestion
For the path lookup, Grafana uses JSON obtained from querying GitLab's API [`/api/v4/user`](https://docs.gitlab.com/ee/api/users.html#list-current-user-for-normal-users) endpoint. The result of evaluating the `role_attribute_path` JMESPath expression must be a valid Grafana role, for example, `Viewer`, `Editor` or `Admin`. For more information about roles and permissions in Grafana, refer to [Organization roles]({{< relref "../permissions/organization_roles.md" >}}).
```

GrafanaWriter commented on docs/sources/auth/gitlab.md:
````diff
@@ -95,21 +95,36 @@ characters. Make sure you always use the group or subgroup name as it appears
+role_attribute_path = is_admin && 'Admin' || 'Viewer'
+```
+
+This would allow you every GitLab Admin to be an Admin in Grafana.
````
```suggestion
This allows every GitLab Admin to be an Admin in Grafana.
```

GrafanaWriter reviewed (commented):
Thank you!

henrysachs committed:
Apply suggestions from code review

Co-authored-by: Fiona Artiaga <89225282+GrafanaWriter@users.noreply.github.com>

henrysachs commented:
All valid suggestions, as I'm not a native speaker. :D

achatterjee-grafana reviewed (commented):
Looks good. Added minor copy esit suggestions.

henrysachs committed:
docs: add suggestions from tech writers

henrysachs commented:
so all suggestions are committed.

achatterjee-grafana commented:
@splattael, thank you so much for contributing!

morsik commented:
Hello, is this feature still exists or it was removed?

I'm trying to configure this seemingly easy feature, and I simply can't make `is_admin && "GrafanaAdmin"` to work... and I just found out that [this PR](https://github.com/grafana/grafana/pull/71834) removed documentation about great `is_admin` feature that from this PR.
